x/programs/v2/vm/actions/program_execute.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/ava-labs/hypersdk/x/programs/v2/vm/storage"
+)
+
+var _ chain.Action = (*ProgramExecute)(nil)
+
+type ProgramExecute struct {
+	Function string      `json:"programFunction"`
+	MaxUnits uint64      `json:"maxUnits"`
+	Params   []CallParam `json:"params"`
+
+	Log logging.Logger
+}
+
+type ProgramStore struct {
+	state.Mutable
+}
+
+func (s *ProgramStore) GetProgramBytes(ctx context.Context, programID ids.ID) ([]byte, error) {
+	// TODO: take fee out of balance?
+	programBytes, exists, err := storage.GetProgram(ctx, s, programID)
+	if err != nil {
+		return []byte{}, err
+	}
+	if !exists {
+		return []byte{}, errors.New("unknown program")
+	}
+
+	return programBytes, nil
+}
+
+func (*ProgramExecute) GetTypeID() uint8 {
+	return programExecuteID
+}
+
+func (*ProgramExecute) StateKeys(_ codec.Address, _ ids.ID) state.Keys {
+	return state.Keys{}
+}
+
+func (*ProgramExecute) StateKeysMaxChunks() []uint16 {
+	return []uint16{storage.ProgramChunks}
+}
+
+func (t *ProgramExecute) Execute(
+	ctx context.Context,
+	_ chain.Rules,
+	mu state.Mutable,
+	_ int64,
+	_ codec.Address,
+	_ ids.ID,
+) (success bool, computeUnits uint64, output []byte, err error) {
+	if len(t.Function) == 0 {
+		return false, 1, OutputValueZero, nil
+	}
+	if len(t.Params) == 0 {
+		return false, 1, OutputValueZero, nil
+	}
+
+	programID, ok := t.Params[0].Value.(ids.ID)
+	if !ok {
+		return false, 1, utils.ErrBytes(errors.New("invalid call param: must be ID")), nil
+	}
+
+	params, err := SerializeParams(t.Params[1:])
+	if err != nil {
+		return false, 0, []byte{}, err
+	}
+
+	cfg := runtime.NewConfig()
+	store := &ProgramStore{
+		Mutable: mu,
+	}
+	rt := runtime.NewRuntime(cfg, t.Log, store)
+	callInfo := &runtime.CallInfo{
+		State:        mu,
+		Actor:        ids.Empty,
+		Account:      ids.Empty,
+		ProgramID:    programID,
+		Fuel:         t.MaxUnits,
+		FunctionName: t.Function,
+		Params:       params,
+	}
+	output, err = rt.CallProgram(ctx, callInfo)
+	if err != nil {
+		return false, 0, output, err
+	}
+	// TODO don't exhaust all fuel here
+	return true, 0, output, nil
+}
+
+func (*ProgramExecute) MaxComputeUnits(chain.Rules) uint64 {
+	return ProgramExecuteComputeUnits
+}
+
+func (*ProgramExecute) Size() int {
+	return ed25519.PublicKeyLen + consts.Uint64Len
+}
+
+func (*ProgramExecute) Marshal(_ *codec.Packer) {
+	// TODO
+}
+
+func (*ProgramExecute) GetBalance() (uint64, error) {
+	// TODO implement metering once available
+	// return t.rt.Meter().GetBalance()
+	return 0, nil
+}
+
+func UnmarshalProgramExecute(_ *codec.Packer) (chain.Action, error) {
+	// TODO
+	return nil, nil
+}
+
+func (*ProgramExecute) ValidRange(chain.Rules) (int64, int64) {
+	// Returning -1, -1 means that the action is always valid.
+	return -1, -1
+}
+
+// CallParam defines a value to be passed to a guest function.
+type CallParam struct {
+	Value interface{} `json,yaml:"value"`
+}
+
+func SerializeParams(p []CallParam) ([]byte, error) {
```

Comment:
@dboehm-avalabs, we've talked about this, but I don't remember what the conclusion was. *I think* we said that it would make sense for the user to serialize the params before constructing the action, right?

Reply:
I agree, though we do need some preprocessing of the params somewhere.  We currently support address aliasing in the simulator steps, so we need to process the params as a list *before* they get serialized.